File: ews_analysis.py
"""
Early Warning System (EWS) Score Analyzer
==========================================
Processes patient-level EWS scores to:
  1. Compute the highest EWS score per patient
  2. Plot the distribution of highest scores
  3. Compute diagnostic accuracy metrics at a user-defined cutoff threshold

Input DataFrame format:
  - One row per patient
  - Multiple columns of EWS scores (e.g., score_1, score_2, ..., score_N)
  - One binary outcome column:
        1 = survived and discharged
        0 = died in hospital

Usage:
    results = analyze_ews(df, score_cols, outcome_col, threshold=60)
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
from scipy import stats


# ---------------------------------------------------------------------------
# Core Analysis Function
# ---------------------------------------------------------------------------

def analyze_ews(
    df: pd.DataFrame,
    score_cols: list[str],
    outcome_col: str,
    threshold: float = 60,
    plot: bool = True,
    group_col: str = None,
) -> dict:
    """
    Analyze EWS scores and compute diagnostic accuracy metrics.

    Parameters
    ----------
    df : pd.DataFrame
        Patient-level data. One row per patient.
    score_cols : list[str]
        Column names containing EWS scores (one column per time point).
        Alternatively, pass a single column name if highest score is pre-computed.
    outcome_col : str
        Binary outcome column. 1 = survived/discharged, 0 = died in hospital.
    threshold : float
        EWS cutoff. Patients with highest score >= threshold are flagged as
        high-risk (predicted to die). Default = 60.
    plot : bool
        Whether to generate and display distribution plots. Default = True.
    group_col : str, optional
        Column name for subgroup analysis (e.g., 'treatment_group').
        If provided, metrics are computed per group in addition to overall.

    Returns
    -------
    dict with keys:
        'highest_scores'  : pd.Series of highest EWS score per patient
        'metrics'         : dict of overall diagnostic accuracy metrics
        'group_metrics'   : dict of per-group metrics (if group_col provided)
        'confusion_matrix': 2x2 array [[TN, FP], [FN, TP]]
    """

    # --- 1. Compute highest score per patient --------------------------------
    if isinstance(score_cols, str):
        score_cols = [score_cols]

    df = df.copy()
    df["_highest_score"] = df[score_cols].max(axis=1)
    df["_outcome"] = df[outcome_col]

    highest_scores = df["_highest_score"]

    # --- 2. Apply threshold to get predictions --------------------------------
    # Prediction: 1 = predicted death (score >= threshold), 0 = predicted survival
    df["_predicted_death"] = (df["_highest_score"] >= threshold).astype(int)

    # Actual death: outcome == 0
    df["_actual_death"] = (df["_outcome"] == 0).astype(int)

    # --- 3. Compute overall metrics -------------------------------------------
    overall_metrics = _compute_metrics(df["_actual_death"], df["_predicted_death"], threshold)

    # --- 4. Per-group metrics (optional) --------------------------------------
    group_metrics = {}
    if group_col is not None:
        for grp, grp_df in df.groupby(group_col):
            group_metrics[grp] = _compute_metrics(
                grp_df["_actual_death"], grp_df["_predicted_death"], threshold
            )

    # --- 5. Confusion matrix --------------------------------------------------
    tn = int(((df["_actual_death"] == 0) & (df["_predicted_death"] == 0)).sum())
    fp = int(((df["_actual_death"] == 0) & (df["_predicted_death"] == 1)).sum())
    fn = int(((df["_actual_death"] == 1) & (df["_predicted_death"] == 0)).sum())
    tp = int(((df["_actual_death"] == 1) & (df["_predicted_death"] == 1)).sum())
    cm = np.array([[tn, fp], [fn, tp]])

    # --- 6. Print summary -----------------------------------------------------
    _print_summary(overall_metrics, group_metrics, cm, threshold)

    # --- 7. Plot --------------------------------------------------------------
    if plot:
        _plot_distribution(df, highest_scores, threshold, group_col)

    return {
        "highest_scores": highest_scores,
        "metrics": overall_metrics,
        "group_metrics": group_metrics,
        "confusion_matrix": cm,
        "data": df,
    }


# ---------------------------------------------------------------------------
# Metric Computation
# ---------------------------------------------------------------------------

def _compute_metrics(actual_death: pd.Series, predicted_death: pd.Series, threshold: float) -> dict:
    """Compute diagnostic accuracy metrics and 95% CIs."""

    n = len(actual_death)
    tp = int(((actual_death == 1) & (predicted_death == 1)).sum())
    tn = int(((actual_death == 0) & (predicted_death == 0)).sum())
    fp = int(((actual_death == 0) & (predicted_death == 1)).sum())
    fn = int(((actual_death == 1) & (predicted_death == 0)).sum())

    sensitivity  = tp / (tp + fn) if (tp + fn) > 0 else np.nan
    specificity  = tn / (tn + fp) if (tn + fp) > 0 else np.nan
    ppv          = tp / (tp + fp) if (tp + fp) > 0 else np.nan
    npv          = tn / (tn + fn) if (tn + fn) > 0 else np.nan
    accuracy     = (tp + tn) / n if n > 0 else np.nan
    prevalence   = (tp + fn) / n if n > 0 else np.nan

    # 95% Wilson confidence intervals for accuracy
    acc_ci = _wilson_ci(tp + tn, n)

    return {
        "n": n,
        "threshold": threshold,
        "prevalence": prevalence,
        "TP": tp, "TN": tn, "FP": fp, "FN": fn,
        "sensitivity": sensitivity,
        "specificity": specificity,
        "PPV": ppv,
        "NPV": npv,
        "accuracy": accuracy,
        "accuracy_95ci": acc_ci,
    }


def _wilson_ci(successes: int, n: int, alpha: float = 0.05) -> tuple[float, float]:
    """Wilson score confidence interval."""
    if n == 0:
        return (np.nan, np.nan)
    z = stats.norm.ppf(1 - alpha / 2)
    p = successes / n
    denom = 1 + z**2 / n
    center = (p + z**2 / (2 * n)) / denom
    margin = (z * np.sqrt(p * (1 - p) / n + z**2 / (4 * n**2))) / denom
    return (max(0, center - margin), min(1, center + margin))


# ---------------------------------------------------------------------------
# Printing
# ---------------------------------------------------------------------------

def _print_summary(metrics: dict, group_metrics: dict, cm: np.ndarray, threshold: float):
    sep = "=" * 60

    print(f"\n{sep}")
    print(f"  EWS DIAGNOSTIC ACCURACY  |  Cutoff Threshold: {threshold}")
    print(sep)
    _print_metrics_block("OVERALL", metrics)

    if group_metrics:
        for grp, gm in group_metrics.items():
            _print_metrics_block(f"GROUP: {grp}", gm)

    print(f"\n  Confusion Matrix (rows=Actual, cols=Predicted)")
    print(f"  {'':20s}  Pred Survived  Pred Died")
    print(f"  {'Actual Survived':20s}  {cm[0,0]:>13,}  {cm[0,1]:>9,}")
    print(f"  {'Actual Died':20s}  {cm[1,0]:>13,}  {cm[1,1]:>9,}")
    print(sep + "\n")


def _print_metrics_block(label: str, m: dict):
    ci = m["accuracy_95ci"]
    print(f"\n  [{label}]")
    print(f"  Patients        : {m['n']:,}")
    print(f"  Prevalence      : {m['prevalence']*100:.2f}%")
    print(f"  Accuracy        : {m['accuracy']*100:.2f}%  (95% CI: {ci[0]*100:.2f}% – {ci[1]*100:.2f}%)")
    print(f"  Sensitivity     : {m['sensitivity']*100:.2f}%")
    print(f"  Specificity     : {m['specificity']*100:.2f}%")
    print(f"  PPV             : {m['PPV']*100:.2f}%")
    print(f"  NPV             : {m['NPV']*100:.2f}%")


# ---------------------------------------------------------------------------
# Plotting
# ---------------------------------------------------------------------------

def _plot_distribution(df: pd.DataFrame, highest_scores: pd.Series, threshold: float, group_col: str):
    """Plot histogram of highest EWS scores, split by outcome and optionally by group."""

    survived = df.loc[df["_actual_death"] == 0, "_highest_score"]
    died     = df.loc[df["_actual_death"] == 1, "_highest_score"]

    n_groups = df[group_col].nunique() if group_col else 1
    n_cols   = 1 + n_groups if group_col else 1
    fig, axes = plt.subplots(1, n_cols, figsize=(7 * n_cols, 5), squeeze=False)

    # Overall plot
    _draw_histogram(axes[0, 0], survived, died, threshold, "Overall")

    # Per-group plots
    if group_col:
        for i, (grp, grp_df) in enumerate(df.groupby(group_col)):
            g_survived = grp_df.loc[grp_df["_actual_death"] == 0, "_highest_score"]
            g_died     = grp_df.loc[grp_df["_actual_death"] == 1, "_highest_score"]
            _draw_histogram(axes[0, i + 1], g_survived, g_died, threshold, f"Group: {grp}")

    plt.suptitle("Distribution of Highest EWS Score per Patient", fontsize=14, fontweight="bold", y=1.02)
    plt.tight_layout()
    plt.savefig("ews_distribution.png", dpi=150, bbox_inches="tight")
    plt.show()
    print("  Plot saved to: ews_distribution.png")


def _draw_histogram(ax, survived: pd.Series, died: pd.Series, threshold: float, title: str):
    bins = np.linspace(0, 100, 41)  # bins of width 2.5 across 0–100

    ax.hist(survived, bins=bins, alpha=0.6, color="steelblue", label=f"Survived (n={len(survived):,})")
    ax.hist(died,     bins=bins, alpha=0.7, color="firebrick", label=f"Died (n={len(died):,})")
    ax.axvline(threshold, color="black", linestyle="--", linewidth=1.5, label=f"Threshold = {threshold}")

    ax.set_xlabel("Highest EWS Score", fontsize=11)
    ax.set_ylabel("Number of Patients", fontsize=11)
    ax.set_title(title, fontsize=12)
    ax.legend(fontsize=9)
    ax.set_xlim(0, 100)


# ---------------------------------------------------------------------------
# Demo / Example Usage
# ---------------------------------------------------------------------------

#if __name__ == "__main__":
#
    # --- Generate synthetic data resembling the abstract's cohort ------------
#    np.random.seed(42)
#    n_medical  = 500
#    n_surgical = 300

#    def simulate_patients(n, death_rate, high_score_if_died_mean, high_score_if_survived_mean, group_label):
#        n_died     = int(n * death_rate)
#        n_survived = n - n_died

        # Simulate 10 EWS score readings per patient (every 15 min over ~2.5 hrs sample)
        # In practice these would be all readings during the full hospitalization
#        def make_scores(n_pts, peak_mean, peak_std=15):
#            peaks = np.clip(np.random.normal(peak_mean, peak_std, n_pts), 0, 100)
#            scores = []
#            for peak in peaks:
                # Other readings are lower than the peak
#                readings = np.clip(np.random.normal(peak * 0.7, 10, 9), 0, 100)
#                readings = np.append(readings, peak)
#                scores.append(readings)
#            return np.array(scores)

#        scores_died     = make_scores(max(n_died, 1),     high_score_if_died_mean)
#        scores_survived = make_scores(n_survived, high_score_if_survived_mean)

#        score_cols = [f"ews_{i}" for i in range(10)]
#        df_died     = pd.DataFrame(scores_died[:n_died] if n_died > 0 else np.empty((0, 10)), columns=score_cols)
#        df_survived = pd.DataFrame(scores_survived, columns=score_cols)

#        df_died["outcome"]     = 0  # died
#        df_survived["outcome"] = 1  # survived

#        df = pd.concat([df_died, df_survived], ignore_index=True)
#        df["group"] = group_label
#        return df, score_cols

#    df_med, score_cols = simulate_patients(n_medical,  death_rate=0.0086, high_score_if_died_mean=75, high_score_if_survived_mean=35, group_label="Medical")
#    df_surg, _        = simulate_patients(n_surgical,  death_rate=0.001,  high_score_if_died_mean=65, high_score_if_survived_mean=38, group_label="Surgical")

#   df_all = pd.concat([df_med, df_surg], ignore_index=True)

#   print("Sample data (first 5 rows):")
#   print(df_all[score_cols[:3] + ["outcome", "group"]].head())

    # --- Run analysis ---------------------------------------------------------
#   results = analyze_ews(
#       df=df_all,
#       score_cols=score_cols,
#       outcome_col="outcome",
#       threshold=60,
#       plot=True,
#       group_col="group",
#   )

#---------------
# Read input data from sample data file
#---------------

import pandas as pd
from ews_analysis import analyze_ews

df = pd.read_csv("ews_sample_data.csv")
score_cols = [c for c in df.columns if c.startswith("ews_")]

results = analyze_ews(df, score_cols=score_cols, outcome_col="outcome", threshold=60, group_col="group")

File: ews_analysis.js
/**
 * ews_analysis.js
 * ===============
 * Early Warning System (EWS) Diagnostic Accuracy Analyzer
 *
 * Processes patient-level EWS score data to:
 *   1. Compute the highest EWS score per patient across all time-point columns
 *   2. Apply a user-defined threshold to classify patients as high-risk or low-risk
 *   3. Compute diagnostic accuracy metrics (Sensitivity, Specificity, PPV, NPV,
 *      Accuracy with 95% Wilson CI) against a binary mortality outcome
 *   4. Optionally stratify all metrics by a subgroup column
 *
 * Input data format:
 *   An array of plain objects (e.g. parsed from CSV), one object per patient.
 *   Each object should have:
 *     - One or more numeric EWS score fields (e.g. ews_001 … ews_096)
 *     - A binary outcome field: 1 = survived/discharged, 0 = died in hospital
 *     - (Optional) a group field for subgroup analysis (e.g. "Medical", "Surgical")
 *
 * Usage (Node.js):
 *   const { analyzeEWS } = require('./ews_analysis');
 *
 *   const results = analyzeEWS({
 *     data        : rows,          // array of patient objects
 *     scoreCols   : ['ews_001', 'ews_002', ...],  // EWS score column names
 *     outcomeCol  : 'outcome',     // binary outcome column (1=survived, 0=died)
 *     threshold   : 60,            // risk cutoff (default: 60)
 *     groupCol    : 'group',       // optional subgroup column
 *     verbose     : true           // print summary to console (default: true)
 *   });
 *
 * Returns:
 *   {
 *     patients     : [...],   // enriched patient array with highestScore added
 *     metrics      : {...},   // overall diagnostic metrics
 *     groupMetrics : {...},   // per-group metrics (if groupCol provided)
 *   }
 *
 * Compatible with Node.js (CommonJS & ESM) and modern browsers.
 * No external dependencies required.
 */

'use strict';

// ─────────────────────────────────────────────────────────────────────────────
// STATISTICS UTILITIES
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Approximation of the standard normal inverse CDF (probit function).
 * Uses the Abramowitz & Stegun rational approximation (formula 26.2.17).
 * Accurate to ~3×10⁻³ across the full range.
 *
 * @param {number} p - Probability (0 < p < 1)
 * @returns {number} z-score corresponding to p
 */
function normInvCDF(p) {
  if (p <= 0 || p >= 1) throw new RangeError('p must be in (0, 1)');
  const a = [2.515517, 0.802853, 0.010328];
  const b = [1.432788, 0.189269, 0.001308];
  const sign = p < 0.5 ? -1 : 1;
  const pp   = p < 0.5 ? p : 1 - p;
  const t    = Math.sqrt(-2 * Math.log(pp));
  const num  = a[0] + a[1] * t + a[2] * t * t;
  const den  = 1 + b[0] * t + b[1] * t * t + b[2] * t * t * t;
  return sign * (t - num / den);
}

/**
 * Wilson score confidence interval for a proportion.
 *
 * @param {number} successes - Number of successes (e.g. correct predictions)
 * @param {number} n         - Total observations
 * @param {number} [alpha]   - Significance level (default 0.05 → 95% CI)
 * @returns {{ lower: number, upper: number }}
 */
function wilsonCI(successes, n, alpha = 0.05) {
  if (n === 0) return { lower: NaN, upper: NaN };
  const z      = normInvCDF(1 - alpha / 2);
  const p      = successes / n;
  const denom  = 1 + (z * z) / n;
  const center = (p + (z * z) / (2 * n)) / denom;
  const margin = (z * Math.sqrt(p * (1 - p) / n + (z * z) / (4 * n * n))) / denom;
  return {
    lower: Math.max(0, center - margin),
    upper: Math.min(1, center + margin),
  };
}


// ─────────────────────────────────────────────────────────────────────────────
// CORE METRIC COMPUTATION
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Compute diagnostic accuracy metrics for a set of enriched patient objects.
 *
 * @param {object[]} patients  - Array of patient objects, each with:
 *                               { highestScore: number, outcome: number (0|1) }
 * @param {number}   threshold - Score cutoff; score >= threshold → predicted death
 * @returns {object} Metrics object (see below)
 */
function computeMetrics(patients, threshold) {
  let tp = 0, tn = 0, fp = 0, fn = 0;

  for (const p of patients) {
    const predictedDeath = p.highestScore >= threshold;
    const actualDeath    = p.outcome === 0;

    if ( actualDeath &&  predictedDeath) tp++;
    if (!actualDeath && !predictedDeath) tn++;
    if (!actualDeath &&  predictedDeath) fp++;
    if ( actualDeath && !predictedDeath) fn++;
  }

  const n           = patients.length;
  const sensitivity = (tp + fn) > 0 ? tp / (tp + fn) : NaN;
  const specificity = (tn + fp) > 0 ? tn / (tn + fp) : NaN;
  const ppv         = (tp + fp) > 0 ? tp / (tp + fp) : NaN;
  const npv         = (tn + fn) > 0 ? tn / (tn + fn) : NaN;
  const accuracy    = n > 0 ? (tp + tn) / n : NaN;
  const prevalence  = n > 0 ? (tp + fn) / n : NaN;
  const accuracyCI  = wilsonCI(tp + tn, n);

  return {
    n,
    threshold,
    prevalence,
    tp, tn, fp, fn,
    sensitivity,
    specificity,
    ppv,
    npv,
    accuracy,
    accuracyCI,             // { lower, upper }
    confusionMatrix: {      // 2×2 as nested object for clarity
      actualSurvived: { predictedSurvived: tn, predictedDied: fp },
      actualDied:     { predictedSurvived: fn, predictedDied: tp },
    },
  };
}


// ─────────────────────────────────────────────────────────────────────────────
// HIGHEST SCORE EXTRACTION
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Enrich each patient row with their highest EWS score across all score columns.
 *
 * @param {object[]} data       - Raw patient rows (array of plain objects)
 * @param {string[]} scoreCols  - Column names to scan for EWS scores
 * @param {string}   outcomeCol - Column name for binary outcome (1=survived, 0=died)
 * @returns {object[]} Enriched patient array; invalid rows are filtered out
 */
function extractHighestScores(data, scoreCols, outcomeCol) {
  const enriched = [];

  for (const row of data) {
    const scores = scoreCols
      .map(col => parseFloat(row[col]))
      .filter(v => !isNaN(v));

    if (scores.length === 0) continue;         // skip rows with no valid scores

    const outcome = parseInt(row[outcomeCol], 10);
    if (isNaN(outcome)) continue;              // skip rows with missing outcome

    enriched.push({
      ...row,                                  // preserve all original fields
      outcome,                                 // store as parsed integer (overrides string)
      highestScore: Math.max(...scores),
    });
  }

  return enriched;
}


// ─────────────────────────────────────────────────────────────────────────────
// SCORE DISTRIBUTION SUMMARY
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Compute a histogram of highest EWS scores (binned 0–100 in steps of 2.5).
 * Returned as an array of { binStart, binEnd, countSurvived, countDied, countTotal }.
 *
 * @param {object[]} patients - Enriched patient array
 * @param {number}   [bins]   - Number of bins (default 40 → width 2.5)
 * @returns {object[]} Histogram bins
 */
function scoreHistogram(patients, bins = 40) {
  const width  = 100 / bins;
  const counts = Array.from({ length: bins }, (_, i) => ({
    binStart:      parseFloat((i * width).toFixed(4)),
    binEnd:        parseFloat(((i + 1) * width).toFixed(4)),
    countSurvived: 0,
    countDied:     0,
    countTotal:    0,
  }));

  for (const p of patients) {
    const idx = Math.min(Math.floor(p.highestScore / width), bins - 1);
    counts[idx].countTotal++;
    if (p.outcome === 1) counts[idx].countSurvived++;
    else                 counts[idx].countDied++;
  }

  return counts;
}


// ─────────────────────────────────────────────────────────────────────────────
// CONSOLE SUMMARY PRINTER
// ─────────────────────────────────────────────────────────────────────────────

function pct(v, dec = 2) {
  return isNaN(v) ? 'N/A' : (v * 100).toFixed(dec) + '%';
}

function printMetricsBlock(label, m) {
  const ci = m.accuracyCI;
  const ciStr = (isNaN(ci.lower) || isNaN(ci.upper))
    ? 'N/A'
    : `${pct(ci.lower)} – ${pct(ci.upper)}`;

  console.log(`\n  [${label}]`);
  console.log(`  Patients      : ${m.n.toLocaleString()}`);
  console.log(`  Prevalence    : ${pct(m.prevalence)}`);
  console.log(`  Accuracy      : ${pct(m.accuracy)}  (95% CI: ${ciStr})`);
  console.log(`  Sensitivity   : ${pct(m.sensitivity)}`);
  console.log(`  Specificity   : ${pct(m.specificity)}`);
  console.log(`  PPV           : ${pct(m.ppv)}`);
  console.log(`  NPV           : ${pct(m.npv)}`);
  console.log(`  TP / TN / FP / FN : ${m.tp} / ${m.tn} / ${m.fp} / ${m.fn}`);
}

function printSummary(metrics, groupMetrics, threshold) {
  const sep = '='.repeat(62);
  console.log(`\n${sep}`);
  console.log(`  EWS DIAGNOSTIC ACCURACY  |  Threshold: ${threshold}`);
  console.log(sep);

  printMetricsBlock('OVERALL', metrics);

  for (const [grp, gm] of Object.entries(groupMetrics)) {
    printMetricsBlock(`GROUP: ${grp}`, gm);
  }

  const cm = metrics.confusionMatrix;
  console.log('\n  Confusion Matrix (Overall)');
  console.log('                       Pred Survived   Pred Died');
  console.log(`  Actual Survived  :   ${String(cm.actualSurvived.predictedSurvived).padStart(13)}   ${String(cm.actualSurvived.predictedDied).padStart(9)}`);
  console.log(`  Actual Died      :   ${String(cm.actualDied.predictedSurvived).padStart(13)}   ${String(cm.actualDied.predictedDied).padStart(9)}`);
  console.log(`${sep}\n`);
}


// ─────────────────────────────────────────────────────────────────────────────
// MAIN EXPORTED FUNCTION
// ─────────────────────────────────────────────────────────────────────────────

/**
 * analyzeEWS — Main entry point.
 *
 * @param {object}   options
 * @param {object[]} options.data        - Array of patient row objects
 * @param {string[]} options.scoreCols   - EWS score column names
 * @param {string}   options.outcomeCol  - Binary outcome column (1=survived, 0=died)
 * @param {number}   [options.threshold] - Risk cutoff score (default: 60)
 * @param {string}   [options.groupCol]  - Optional subgroup column name
 * @param {boolean}  [options.verbose]   - Print summary to console (default: true)
 *
 * @returns {{
 *   patients     : object[],   // enriched patient array with highestScore
 *   metrics      : object,     // overall diagnostic metrics
 *   groupMetrics : object,     // per-group metrics keyed by group value
 *   histogram    : object[],   // overall score histogram bins
 * }}
 */
function analyzeEWS({
  data,
  scoreCols,
  outcomeCol,
  threshold = 60,
  groupCol  = null,
  verbose   = true,
}) {
  // ── Validate inputs ──────────────────────────────────────────────────────
  if (!Array.isArray(data) || data.length === 0) {
    throw new Error('analyzeEWS: "data" must be a non-empty array of patient objects.');
  }
  if (!Array.isArray(scoreCols) || scoreCols.length === 0) {
    throw new Error('analyzeEWS: "scoreCols" must be a non-empty array of column name strings.');
  }
  if (typeof outcomeCol !== 'string' || !outcomeCol) {
    throw new Error('analyzeEWS: "outcomeCol" must be a non-empty string.');
  }
  if (typeof threshold !== 'number' || threshold < 0 || threshold > 100) {
    throw new Error('analyzeEWS: "threshold" must be a number between 0 and 100.');
  }

  // ── Step 1: Extract highest score per patient ────────────────────────────
  const patients = extractHighestScores(data, scoreCols, outcomeCol);

  if (patients.length === 0) {
    throw new Error('analyzeEWS: No valid patient rows after parsing. Check scoreCols and outcomeCol.');
  }

  // ── Step 2: Overall metrics ──────────────────────────────────────────────
  const metrics = computeMetrics(patients, threshold);

  // ── Step 3: Per-group metrics (optional) ─────────────────────────────────
  const groupMetrics = {};
  if (groupCol) {
    const groups = {};
    for (const p of patients) {
      const g = String(p[groupCol] ?? 'Unknown');
      if (!groups[g]) groups[g] = [];
      groups[g].push(p);
    }
    for (const [grp, gPatients] of Object.entries(groups)) {
      groupMetrics[grp] = computeMetrics(gPatients, threshold);
    }
  }

  // ── Step 4: Score distribution histogram ─────────────────────────────────
  const histogram = scoreHistogram(patients);

  // ── Step 5: Console output ───────────────────────────────────────────────
  if (verbose) {
    printSummary(metrics, groupMetrics, threshold);
  }

  return { patients, metrics, groupMetrics, histogram };
}


// ─────────────────────────────────────────────────────────────────────────────
// EXPORTS  (CommonJS + ESM compatible)
// ─────────────────────────────────────────────────────────────────────────────

if (typeof module !== 'undefined' && module.exports) {
  // CommonJS (Node.js require)
  module.exports = { analyzeEWS, computeMetrics, wilsonCI, normInvCDF, scoreHistogram };
} else if (typeof window !== 'undefined') {
  // Browser global
  window.EWSAnalysis = { analyzeEWS, computeMetrics, wilsonCI, normInvCDF, scoreHistogram };
}


// ─────────────────────────────────────────────────────────────────────────────
// DEMO  (runs only when executed directly: node ews_analysis.js)
// ─────────────────────────────────────────────────────────────────────────────

if (typeof require !== 'undefined' && require.main === module) {
  const fs  = require('fs');
  const path = require('path');

  // Try to load the sample CSV if available
  const csvPath = path.join(__dirname, 'ews_sample_data.csv');

  if (fs.existsSync(csvPath)) {
    console.log(`\nLoading sample data from: ${csvPath}`);

    const raw  = fs.readFileSync(csvPath, 'utf8');
    const lines = raw.trim().split('\n');
    const headers = lines[0].split(',').map(h => h.trim().replace(/^"|"$/g, ''));

    function parseCSVLine(line) {
      const fields = [];
      let current = '', inQuotes = false;
      for (let i = 0; i < line.length; i++) {
        const ch = line[i];
        if (ch === '"') { inQuotes = !inQuotes; }
        else if (ch === ',' && !inQuotes) { fields.push(current); current = ''; }
        else { current += ch; }
      }
      fields.push(current);
      return fields;
    }

    const rows = lines.slice(1).map(line => {
      const values = parseCSVLine(line);
      const obj = {};
      headers.forEach((h, i) => { obj[h] = (values[i] ?? '').trim(); });
      return obj;
    });

    const scoreCols = headers.filter(h => h.startsWith('ews_'));
    console.log(`Patients: ${rows.length.toLocaleString()}  |  Score columns: ${scoreCols.length}`);

    const results = analyzeEWS({
      data       : rows,
      scoreCols,
      outcomeCol : 'outcome',
      threshold  : 60,
      groupCol   : 'group',
      verbose    : true,
    });

    // Show a sample of the histogram
    console.log('Score Distribution (bins with patients):');
    results.histogram
      .filter(b => b.countTotal > 0)
      .forEach(b => {
        const bar = '█'.repeat(Math.ceil(b.countTotal / 5));
        console.log(
          `  ${String(b.binStart.toFixed(1)).padStart(5)}–${String(b.binEnd.toFixed(1)).padEnd(5)} ` +
          `| ${String(b.countTotal).padStart(4)} patients  ${bar}`
        );
      });

  } else {
    // Minimal inline demo if no CSV present
    console.log('\nNo CSV found — running inline demo with synthetic data.\n');

    const demoData = [
      { ews_1: '72', ews_2: '80', ews_3: '65', outcome: '0', group: 'Medical'  },
      { ews_1: '20', ews_2: '35', ews_3: '28', outcome: '1', group: 'Medical'  },
      { ews_1: '45', ews_2: '55', ews_3: '62', outcome: '1', group: 'Medical'  },
      { ews_1: '88', ews_2: '91', ews_3: '70', outcome: '0', group: 'Surgical' },
      { ews_1: '15', ews_2: '22', ews_3: '18', outcome: '1', group: 'Surgical' },
      { ews_1: '60', ews_2: '58', ews_3: '63', outcome: '1', group: 'Surgical' },
    ];

    analyzeEWS({
      data      : demoData,
      scoreCols : ['ews_1', 'ews_2', 'ews_3'],
      outcomeCol: 'outcome',
      threshold : 60,
      groupCol  : 'group',
      verbose   : true,
    });
  }
}
